<a href="https://colab.research.google.com/github/arqPy/CAD/blob/main/A_Nodal_Colab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Análise Nodal


LKT

$AI = 0$



In [4]:
def nos_ramos(circuito):

  nR = len(circuito)
  listNodes= set([])
  for ramo in circuito:
    no1 = ramo['NP']
    no2 = ramo['NC']

    listNodes.add(no1)
    listNodes.add(no2)
  nN = len(listNodes)
  return (nN - 1, nR)



In [6]:
# PROGRAMA PRINCIPAL --- Lê o grafo em um dicionário python
import numpy as np
def executaCircuito(circuito):
  (nN, nR) = nos_ramos(circuito)
  A = np.zeros((nN, nR))

  G = np.zeros((nR, nR))
  Is = np.zeros((nR, 1))
  Vs = np.zeros((nR, 1))

  for ramo in circuito:
    no1 = ramo['NP']
    no2 = ramo['NC']

    i = ramo['id']
    print(no1, no2, i)
    G[i - 1, i - 1] = 1./ramo['R']
    Vs[i - 1, 0] = ramo['Vs']
    Is[i - 1, 0] = ramo['Is']

    if (no1 != 0):
      A[no1-1, i-1 ] = 1.
    if (no2 != 0):
      A[no2-1, i-1 ] = -1.

  # Matrizes
  A = np.matrix(A, dtype=np.float64)
  G = np.matrix(G, dtype=np.float64)
  Vs = np.matrix(Vs, dtype=np.float64)
  Is = np.matrix(Is, dtype=np.float64)

  print(100*"*")
  print("Entrada ")
  print("\nA \n", A)
  print("\nG \n", G)
  print("\nVs \n", Vs)
  print("\nIs \n", Is)

  Gb = A*G*A.T
  In = A*(G*Vs - Is)
  E = (Gb.I)*In
  V = A.T*E
  I = Is + G*(V - Vs)

  print(100*"*")
  print("\nSaída ")
  print("\nCondutâncias de barras \n", Gb)
  print("\nCorrente Independente Resultante no Nó \n", In)
  print("\nTensão de Nó \n", E)
  print("\nTensão de ramo \n", V)
  print("\nCorrente de ramo \n", I)
  return

1 - *Exemplo*

![](https://github.com/arqPy/CAD/blob/main/Circuito_eletrico/fig/Circuito-Grafo-01.png?raw=true)

In [7]:
# grafo do circuito TOPOLOGIA - LISTA DE RAMOS
circuito = [
    {'id': 1, 'NP': 0, 'NC': 1, 'R': 7, 'Vs': 0, 'Is': 16},
    {'id': 2, 'NP': 1, 'NC': 2, 'R': 1e-7, 'Vs': -2, 'Is': 0},
    {'id': 3, 'NP': 2, 'NC': 3, 'R': 2, 'Vs': 0, 'Is': 0},
    {'id': 4, 'NP': 2, 'NC': 0, 'R': 6, 'Vs': 0, 'Is': 0},
    {'id': 5, 'NP': 3, 'NC': 0, 'R': 4, 'Vs': 8, 'Is': 0},
    {'id': 6, 'NP': 0, 'NC': 3, 'R': 2, 'Vs': -6, 'Is': 0},
]
nos_ramos(circuito)

(3, 6)

In [8]:
executaCircuito(circuito)

0 1 1
1 2 2
2 3 3
2 0 4
3 0 5
0 3 6
****************************************************************************************************
Entrada 

A 
 [[-1.  1.  0.  0.  0.  0.]
 [ 0. -1.  1.  1.  0.  0.]
 [ 0.  0. -1.  0.  1. -1.]]

G 
 [[1.42857143e-01 0.00000000e+00 0.00000000e+00 0.00000000e+00
  0.00000000e+00 0.00000000e+00]
 [0.00000000e+00 1.00000000e+07 0.00000000e+00 0.00000000e+00
  0.00000000e+00 0.00000000e+00]
 [0.00000000e+00 0.00000000e+00 5.00000000e-01 0.00000000e+00
  0.00000000e+00 0.00000000e+00]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 1.66666667e-01
  0.00000000e+00 0.00000000e+00]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00
  2.50000000e-01 0.00000000e+00]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00
  0.00000000e+00 5.00000000e-01]]

Vs 
 [[ 0.]
 [-2.]
 [ 0.]
 [ 0.]
 [ 8.]
 [-6.]]

Is 
 [[16.]
 [ 0.]
 [ 0.]
 [ 0.]
 [ 0.]
 [ 0.]]
*********************************************************************************************

ENTRE COM O GRAFO REFERENTE AO CIRCUITO ABAIXO

![](https://github.com/arqPy/CAD/blob/main/Circuito_eletrico/fig/CIRCUITO_REDUZIDO.jpeg?raw=true)





In [10]:
# Reduzindo
circuito = [
    {'id': 1, 'NP': 0, 'NC': 1, 'R': 7, 'Vs': 0, 'Is': 16},
    {'id': 2, 'NP': 1, 'NC': 2, 'R': 1e-7, 'Vs': -2, 'Is': 0},
    {'id': 3, 'NP': 2, 'NC': 0, 'R': 6, 'Vs': 0, 'Is': 0},
    {'id': 4, 'NP': 0, 'NC': 2, 'R': 10/3, 'Vs': -20/3, 'Is': 0}
]
executaCircuito(circuito)

0 1 1
1 2 2
2 0 3
0 2 4
****************************************************************************************************
Entrada 

A 
 [[-1.  1.  0.  0.]
 [ 0. -1.  1. -1.]]

G 
 [[1.42857143e-01 0.00000000e+00 0.00000000e+00 0.00000000e+00]
 [0.00000000e+00 1.00000000e+07 0.00000000e+00 0.00000000e+00]
 [0.00000000e+00 0.00000000e+00 1.66666667e-01 0.00000000e+00]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 3.00000000e-01]]

Vs 
 [[ 0.        ]
 [-2.        ]
 [ 0.        ]
 [-6.66666667]]

Is 
 [[16.]
 [ 0.]
 [ 0.]
 [ 0.]]
****************************************************************************************************

Saída 

Condutâncias de barras 
 [[ 10000000.14285714 -10000000.        ]
 [-10000000.          10000000.46666667]]

Corrente Independente Resultante no Nó 
 [[-19999984.]
 [ 20000002.]]

Tensão de Nó 
 [[28.00000087]
 [29.99999967]]

Tensão de ramo 
 [[-28.00000087]
 [ -1.9999988 ]
 [ 29.99999967]
 [-29.99999967]]

Corrente de ramo 
 [[11.99999988]
 [12.038

Circuito comparado com O LTSpice
![](https://moodle.ifg.edu.br/pluginfile.php/1475067/mod_label/intro/anNodal.png)

In [11]:
circuito = [
    {'id': 1, 'NP': 3, 'NC': 1, 'R': 10/7, 'Vs': 0, 'Is': 1},
    {'id': 2, 'NP': 1, 'NC': 2, 'R': 5, 'Vs': 0, 'Is': -2},
    {'id': 3, 'NP': 2, 'NC': 3, 'R': 10, 'Vs': 0, 'Is': 0},
    {'id': 4, 'NP': 0, 'NC': 2, 'R': 4, 'Vs': 0, 'Is': 3},
    {'id': 5, 'NP': 0, 'NC': 3, 'R': 2, 'Vs': 0, 'Is': 0}
]
executaCircuito(circuito)

3 1 1
1 2 2
2 3 3
0 2 4
0 3 5
****************************************************************************************************
Entrada 

A 
 [[-1.  1.  0.  0.  0.]
 [ 0. -1.  1. -1.  0.]
 [ 1.  0. -1.  0. -1.]]

G 
 [[0.7  0.   0.   0.   0.  ]
 [0.   0.2  0.   0.   0.  ]
 [0.   0.   0.1  0.   0.  ]
 [0.   0.   0.   0.25 0.  ]
 [0.   0.   0.   0.   0.5 ]]

Vs 
 [[0.]
 [0.]
 [0.]
 [0.]
 [0.]]

Is 
 [[ 1.]
 [-2.]
 [ 0.]
 [ 3.]
 [ 0.]]
****************************************************************************************************

Saída 

Condutâncias de barras 
 [[ 0.9  -0.2  -0.7 ]
 [-0.2   0.55 -0.1 ]
 [-0.7  -0.1   1.3 ]]

Corrente Independente Resultante no Nó 
 [[ 3.]
 [ 1.]
 [-1.]]

Tensão de Nó 
 [[7.15789474]
 [5.05263158]
 [3.47368421]]

Tensão de ramo 
 [[-3.68421053]
 [ 2.10526316]
 [ 1.57894737]
 [-5.05263158]
 [-3.47368421]]

Corrente de ramo 
 [[-1.57894737]
 [-1.57894737]
 [ 0.15789474]
 [ 1.73684211]
 [-1.73684211]]


V(n001): 7.15789 voltage
V(n003): 3.47368 voltage
V(n002): 5.05263 voltage
I(I3): 3 device_current
I(I2): 2 device_current
I(I1): 1 device_current
I(R6): -1.73684 device_current
I(R5): -0.421053 device_current
I(R4): 1.26316 device_current
I(R3): 0.157895 device_current
I(R2): 1.84211 device_current
I(R1): 0.736842 device_current
    
CORRETO
